tarfs.go

```diff
@@ -100,39 +101,54 @@ func (fs TarWriter) Close() error {
 // a catar).
 type TarReader struct {
 	r *gnutar.Reader
+	root *File
 }
 
-var _ FilesystemReader = TarReader{}
+var _ FilesystemReader = &TarReader{}
 
 // NewTarFS initializes a new instance of a GNU tar archive that can be used
 // for catar archive tar/untar operations.
-func NewTarReader(r io.Reader) TarReader {
-	return TarReader{gnutar.NewReader(r)}
+func NewTarReader(r io.Reader, addRoot bool) *TarReader {
+	var root *File
+	if addRoot {
+		root = &File{
+			Name: ".",
+			Path: ".",
+			Mode: os.ModeDir | 0755,
+		}
+	}
+	return &TarReader{
+		r: gnutar.NewReader(r),
+		root: root,
+	}
 }
 
 // Next returns the next filesystem entry or io.EOF when done. The caller is responsible
 // for closing the returned File object.
-func (fs TarReader) Next() (*File, error) {
-	h, err := fs.r.Next()
-	if err != nil {
-		return nil, err
-	}
-
-	info := h.FileInfo()
-
-	f := &File{
-		Name:       info.Name(),
-		Path:       path.Clean(h.Name),
-		Mode:       info.Mode(),
-		ModTime:    info.ModTime(),
-		Size:       uint64(info.Size()),
-		LinkTarget: h.Linkname,
-		Uid:        h.Uid,
-		Gid:        h.Gid,
-		Xattrs:     h.Xattrs,
-		DevMajor:   uint64(h.Devmajor),
-		DevMinor:   uint64(h.Devminor),
-		Data:       ioutil.NopCloser(fs.r),
+func (fs *TarReader) Next() (f *File, err error) {
+	if fs.root != nil {
+		f = fs.root
+		fs.root = nil
+	} else {
```

Comment:
Might be better to add a return above and remove the `else` to reduce nesting of the block below